src/stan/model/indexing/assign_cl.hpp

```diff
@@ -0,0 +1,369 @@
+ * @throw std::out_of_range If the index is out of bounds.
+ */
+template <typename ExprLhs, typename ScalRhs,
+          require_rev_kernel_expression_t<ExprLhs>* = nullptr,
+          require_arithmetic_t<ScalRhs>* = nullptr>
+inline void assign(ExprLhs&& expr_lhs, const ScalRhs& scal_rhs,
+                   const char* name, const index_uni row_index,
+                   const index_uni col_index) {
+  auto&& lhs_holder = math::block_zero_based(
+      expr_lhs.val_op(), row_index.n_ - 1, col_index.n_ - 1, 1, 1);
+  auto& lhs = lhs_holder;
+  math::arena_matrix_cl<double> prev_val = lhs;
+  lhs = math::constant(scal_rhs, 1, 1);
+  math::reverse_pass_callback(
+      [expr_lhs, row_index, col_index, prev_val]() mutable {
+        auto&& lhs_val = math::block_zero_based(
+            expr_lhs.val_op(), row_index.n_ - 1, col_index.n_ - 1, 1, 1);
+        auto&& lhs_adj = math::block_zero_based(
+            expr_lhs.adj(), row_index.n_ - 1, col_index.n_ - 1, 1, 1);
+        math::results(lhs_adj, lhs_val)
+            = math::expressions(math::constant(0.0, 1, 1), prev_val);
+      });
+}
+
+// assign(rev, rev, ...)
+/**
+ * Assign one reverse mode kernel generator expression to another, using given
+ * index.
+ *
+ * @tparam ExprLhs type of the expression on the left hand side of the
+ * assignment
+ * @tparam ExprRhs type of the expression on the right hand side of the
+ * assignment
+ * @tparam RowIndex type of index
+ * @param[in,out] expr_lhs expression on the left hand side of the assignment
+ * @param expr_rhs expression on the right hand side of the assignment
+ * @param name Name of lvalue variable
+ * @param row_index index used for indexing `expr_lhs`
+ * @throw std::out_of_range If the index is out of bounds.
+ * @throw std::invalid_argument If the right hand side size isn't the same as
+ * the indexed left hand side size.
+ */
+template <typename ExprLhs, typename ExprRhs, typename RowIndex,
+          require_all_rev_kernel_expression_t<ExprLhs, ExprRhs>* = nullptr>
+inline void assign(ExprLhs&& expr_lhs, const ExprRhs& expr_rhs,
+                   const char* name, const RowIndex& row_index) {
+  if (std::is_same<RowIndex, index_omni>::value) {
+    stan::math::check_size_match("omni assign", "left hand side rows",
+                                 expr_lhs.rows(), name, expr_rhs.rows());
+    stan::math::check_size_match("omni assign", "left hand side columns",
+                                 expr_lhs.cols(), name, expr_rhs.cols());
+  }
+  auto&& lhs_holder = rvalue(expr_lhs.val_op(), name, row_index);
+  auto& lhs = lhs_holder;
+  math::arena_matrix_cl<double> prev_vals = lhs;
+  lhs = expr_rhs.val();
+  math::reverse_pass_callback(
+      [expr_lhs, expr_rhs, row_index, name, prev_vals]() mutable {
+        auto&& lhs_val = rvalue(expr_lhs.val_op(), name, row_index);
+        auto&& lhs_adj_holder = rvalue(expr_lhs.adj(), name, row_index);
+        auto& lhs_adj = lhs_adj_holder;
```

Comment:
Why `auto&&` and then `auto&`?

Reply:
Actually this is what `decltype(auto)` is for. I did not think of it when I was first writing this.